mobile/android/android-device-file-system.ts

```diff
@@ -39,16 +44,65 @@ export class AndroidDeviceFileSystem implements Mobile.IDeviceFileSystem {
 					this.adb.executeShellCommand(["chmod", "0777", localToDevicePathData.getDevicePath()]).wait()
 				)
 				.value();
+
+			// Update hashes
+			let deviceHashService = this.getDeviceHashService(deviceAppData.appIdentifier);
+			if (!deviceHashService.updateHashes(localToDevicePaths).wait()) {
+				this.$logger.trace("Unable to find hash file on device. The next livesync command will create it.");
+			}
 		}).future<void>()();
 	}
 
 	public transferDirectory(deviceAppData: Mobile.IDeviceAppData, localToDevicePaths: Mobile.ILocalToDevicePathData[], projectFilesPath: string): IFuture<void> {
 		return (() => {
-			this.adb.executeCommand(["push", projectFilesPath, deviceAppData.deviceProjectRootPath]).wait();
-			let deviceFilePaths = _.map(localToDevicePaths, (localToDevicePathData) => `"${localToDevicePathData.getDevicePath()}"`).join(" ");
+			let devicePaths: string[] = [],
+				currentShasums: string[] = [];
+
+			localToDevicePaths.forEach(localToDevicePathData => {
+				let localPath = localToDevicePathData.getLocalPath();
+				let stats = this.$fs.getFsStats(localPath).wait();
+				if (stats.isFile()) {
+					let fileShasum = this.$fs.getFileShasum(localPath).wait();
+					currentShasums.push(`${localPath} ${fileShasum}`);
+				}
+				devicePaths.push(`"${localToDevicePathData.getDevicePath()}"`);
+			});
+
 			let commandsDeviceFilePath = this.$mobileHelper.buildDevicePath(deviceAppData.deviceProjectRootPath, "nativescript.commands.sh");
-			this.createFileOnDevice(commandsDeviceFilePath, "chmod 0777 " + deviceFilePaths).wait();
-			this.adb.executeShellCommand([commandsDeviceFilePath]).wait();
+
+			let deviceHashService = this.getDeviceHashService(deviceAppData.appIdentifier);
+			let filesToChmodOnDevice: string[] = devicePaths;
+			if (this.$options.force) {
+				this.adb.executeShellCommand(["rm", "-rf", deviceHashService.hashFileDevicePath]).wait();
+				this.adb.executeCommand(["push", projectFilesPath, deviceAppData.deviceProjectRootPath]).wait();
+			} else {
+				// Create or update file hashes on device
+				let oldShasums = deviceHashService.getShasumsFromDevice().wait();
+				if (oldShasums) {
+					let changedShasums = _(currentShasums)
+						.filter(hash => !_.find(oldShasums, h => hash.trim().toLowerCase() === h.trim().toLowerCase()))
```

Comment:
There is no chaining of functions in here. I suggest to use the shorter call

`_.filter(currentShasums, hash => ...);`